### DISCUSSION TITLE
Knitr + pandoc Workflow for Executing Code and Insert Output (R Markdown)

### DISCUSSION THREAD

**Discussion by LucunJi** [Taitava/obsidian-shellcommands#322] [Show and tell]
# Nice workflow, at least for me
[> Code Here <](https://gist.github.com/LucunJi/a8fb6e0a99246cc62664ae5c772965b1)

An competitive alternative for running codes before current extensions (Jupyter/Execute Code) become more mature.

This workflow has 3 parts: one to knit the markdown, one to compile into PDF, and another for opening the PDF.

You can find recommended settings and descriptions for custom variables in that Gist.

# Features
- Very configurable (set the custom variables, configure the defaults file for pandoc, or just edit the code)
- Multiple languages (r, python, as long as knitr supports, but sorry there's no matlab)
- Per-code block settings for inputs, outputs, engines... (see [document of knitr](https://yihui.org/knitr/options/))
- Inserts images and outputs for saving & exporting
- Also supports syntax ```` ```language +x{args}```` because Obsidian does not highlight R Markdown syntax ```` ```{language args}````
- Combining with QuickAdd, we can save file before kintting/exporting

# Possible improvements
- Easier re-running code block: preserve per-block settings, clear the outputs
- Quickly making a code block executable
- Lua pandoc filters for centering images, scaling images in Obsidian's style (`![caption|size](url)`)
- Auto-detect attachment path from `.obsidian/app.json`
- Make an extension for this, as the current code is already growing large

![Peek 2023-02-01 00-23](https://user-images.githubusercontent.com/36262513/215985593-90f06f96-c0a3-41f4-b635-c5455df74ddc.gif)

**Comment by LucunJi** [comment]
Note: run
```bash
mkpipe .Rpipe;
Rscript -e "while (T) {tryCatch(eval(parse(text=readLines('stdin'))), error = function (e) {print(e); sys.calls()})}" < .Rpipe &
```
Makes it possible to use variables across runs bu reusing the same R session with a FIFO.
It's a possible modification to the existing shellcommand.

**Comment by Taitava** [comment]
Nice! 🙂 I'll reply just shortly now. I need to look into this at a better time. Can you elaborate a bit about what is knit? I'm not familiar with R programming language, either, although I've heard the name.

Do you have some practical example use case on which you use this? 🙂 I'm just trying to familiarize myself with this workflow and its possibilities. 🙂

**Comment by LucunJi** [comment]
- RMarkdown _as a format_: just a dialect of markdown, usually has extension `.Rmd` rather than `.md`, allowing to mark code blocks that we want to run.
- RMarkdown _as a system_: a bunch of things for `.Rmd` file, including format specification, runing code, and exporting to PDF.
- RMarkdown _as a R package_: build the `.Rmd` file into a PDF via pandoc and knitr, also supports other formats.
- knitr: an R package for running R code in `.Rmd` and generating a `.md` file with output. You can use it by just calling `knitr::knit('filepath')`. It can do more things such as run code in LaTeX and run Python code.

I actually prefer Python/Matlab over R. But for running some simple code in markdown, RMarkdown seems to be competitive.
